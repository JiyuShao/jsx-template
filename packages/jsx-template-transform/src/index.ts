/*
 * transform jsx code string to js code string
 *
 * @Author: Jiyu Shao
 * @Date: 2019-12-05 16:51:56
 * @Last Modified by: Jiyu Shao
 * @Last Modified time: 2019-12-12 17:51:37
 */
import * as fs from 'fs';
import { transform } from '@babel/core';

interface Options {
  /**
   * @babel/plugin-transform-react-jsx plugin options
   */
  transformOptions?: Record<string, any>;

  /**
   * This prefix will add to code before transformed by babel;
   */
  importPrefix?: string;
}

const jsxTemplateTransform = (filePath: string, options: Options = {}) => {
  const {
    transformOptions = {},
    importPrefix = "import { h } from 'preact';\n",
  } = options;

  const code = fs.readFileSync(filePath, 'utf-8');
  const codeWithPrefix = `${importPrefix}${code}`;

  return transform(codeWithPrefix, {
    filename: filePath,
    presets: [
      [
        '@babel/preset-env',
        {
          modules: 'commonjs',
        },
      ],
    ],
    plugins: [
      [
        '@babel/plugin-transform-react-jsx',
        {
          pragma: 'h', // default pragma is React.createElement
          ...transformOptions,
        },
      ],
      [
        'babel-plugin-transform-imports-with-loader',
        {
          rules: {
            test: /\.jsx/,
            unserializeFunc: 'eval',
            transform: (_: string, filePath: string) => {
              const parsedCode = (jsxTemplateTransform(
                filePath,
                options
              ) as any).code;
              return parsedCode;
            },
          },
        },
      ],
    ],
  });
};

export default jsxTemplateTransform;
