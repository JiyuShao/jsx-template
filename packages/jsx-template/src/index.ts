/*
 * jsx-template modules
 *
 * @Author: Jiyu Shao
 * @Date: 2019-12-05 16:53:38
 * @Last Modified by: Jiyu Shao
 * @Last Modified time: 2019-12-05 17:22:03
 */

import { render as renderToString } from 'preact-render-to-string';
import jsxTemplateTransform from 'jsx-template-transform';

/**
 * compile code into a function
 * @param {string} code jsx code
 * @returns {Function} function that can return html
 */
export const compile = (code: string) => {
  // transform JSX code string
  const transformedCode = jsxTemplateTransform(code);

  // get JS code string
  let finalCode: string;
  if (!transformedCode) {
    throw new Error('jsx-template: compile jsxTemplateTransform code empty');
  } else {
    finalCode = transformedCode.code ? transformedCode.code : '';
  }

  // return function
  return (...args: any[]) => {
    // get codeFunc from JS code string
    const codeFunc = eval(finalCode);

    // exec codeFunc
    const reactElement = codeFunc(...args);
    return renderToString(reactElement);
  };
};
