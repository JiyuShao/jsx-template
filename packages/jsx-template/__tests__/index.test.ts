import * as path from 'path';
import { compileFile } from '../src/index';

describe('`compile` should work with functional component', () => {
  const filename = path.resolve(
    __dirname,
    './fixtures/functional-component.jsx'
  );

  it('compile jsx template to function', () => {
    expect(typeof compileFile(filename) === 'function').toBe(true);
  });

  it('function return html', () => {
    const html = compileFile(filename)();
    expect(html).toMatchSnapshot();
  });
});

describe('`compile` should work with functional component with parameters', () => {
  const filename = path.resolve(
    __dirname,
    './fixtures/functional-component-with-params.jsx'
  );

  it('compile jsx template to function', () => {
    expect(typeof compileFile(filename) === 'function').toBe(true);
  });

  it('function return html', () => {
    const content = 'this is custom content';
    const html = compileFile(filename)(content);
    expect(html.includes(content)).toBe(true);
    expect(html).toMatchSnapshot();
  });
});

describe('`compile` should work with functional component with import', () => {
  const filename = path.resolve(
    __dirname,
    './fixtures/functional-component-with-import.jsx'
  );

  it('compile jsx template to function', () => {
    expect(typeof compileFile(filename) === 'function').toBe(true);
  });

  it('function return html', () => {
    const html = compileFile(filename)();
    expect(html).toMatchSnapshot();
    expect(html.includes('Functional Component')).toBe(true);
  });
});
