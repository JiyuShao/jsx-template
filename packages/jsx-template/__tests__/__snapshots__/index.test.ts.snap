// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`compile\` should work with functional component function return html 1`] = `"<div><h1>Functional Component:</h1><p>Functional Component Content</p></div>"`;

exports[`\`compile\` should work with functional component with import function return html 1`] = `"<div><h1>Functional Component With Import:</h1><div><h1>Functional Component:</h1><p>Functional Component Content</p></div></div>"`;

exports[`\`compile\` should work with functional component with parameters function return html 1`] = `"<div><h1>Functional Component With Params:</h1><p>this is custom content</p></div>"`;
